hack/dev-env/setup-vcluster-env.sh

```diff
@@ -0,0 +1,293 @@
+    if (kubectl api-resources || true) | grep -q "openshift.io"; then
+        OPENSHIFT=true
+    fi
+
+}
+
+
+# wait_for_pods looks all Pods running in k8s context $1, and keeps waiting until running count == $2. 
+wait_for_pods() {
+
+    set +e
+
+    count=0
+    while [ true ]
+    do
+
+        echo "  -> Waiting for $1 pods to be running. Expecting $2 running pods."
+
+        kubectl get pods --context="$1" -A
+        RUNNING_PODS=`kubectl get pods --context="$1" -A | grep "Running" | wc -l | tr -d '[:space:]'`
+
+        if [[ "$RUNNING_PODS" == "$2" ]]; then
+            break
+        fi
+
+        count=$((count+1))
+        if [[ $count -eq 60 ]]; then
+            echo "  -> Timed out waiting for pods to be running."
+            echo "    -> Pods:"
+            kubectl describe pods --context="$1" -n argocd
+            echo "    -> Deployments:"
+            kubectl describe deployments --context="$1" -n argocd
+            echo "    -> Stateful Sets:"
+            kubectl describe statefulsets --context="$1" -n argocd
+            echo "  -> Exiting due to timeout waiting for pods to be running."
+            exit 1
+        fi
+
+        sleep 10
+    done
+
+    echo "  -> Done waiting for $1 pods."
+
+    set -e
+}
+
+
+apply() {
+
+    TMP_DIR=`mktemp -d`
+
+    echo "-> TMP_DIR is $TMP_DIR"
+    cp -r ${SCRIPTPATH}/* $TMP_DIR
+
+    mkdir -p ${TMP_DIR}/argo-cd
+    cd ${TMP_DIR}/argo-cd
+    git init
+    git remote add origin https://github.com/argoproj/argo-cd
+    git fetch --depth=1 origin stable
+    git checkout FETCH_HEAD
+    cd -
+
+    # Comment out 'loadBalancerIP:' lines on OpenShift
+    if [[ "$OPENSHIFT" != "" ]]; then
+        sed -i.bak -e '/loadBalancerIP/s/^/#/' $TMP_DIR/control-plane/redis-service.yaml
+        sed -i.bak -e '/loadBalancerIP/s/^/#/' $TMP_DIR/control-plane/repo-server-service.yaml
+        sed -i.bak -e '/loadBalancerIP/s/^/#/' $TMP_DIR/control-plane/server-service.yaml
+    fi
+
+    LATEST_RELEASE_TAG=`curl -s "https://api.github.com/repos/argoproj/argo-cd/releases/latest" | jq -r .tag_name`
+    sed -i.bak -e "s/LatestReleaseTag/${LATEST_RELEASE_TAG}/" $TMP_DIR/control-plane/kustomization.yaml
+    sed -i.bak -e "s/LatestReleaseTag/${LATEST_RELEASE_TAG}/" $TMP_DIR/agent-autonomous/kustomization.yaml
+    sed -i.bak -e "s/LatestReleaseTag/${LATEST_RELEASE_TAG}/" $TMP_DIR/agent-managed/kustomization.yaml
+
+    # Generate the server secret key for the argocd running on the managed and autonomous agent clusters
+    echo "-> Generate server.secretkey for agent's argocd-secrets"
+    if ! pwmake=$(which pwmake); then
+        pwmake=$(which pwgen)
+    fi
+    echo "data:" >> $TMP_DIR/agent-managed/argocd-secret.yaml
+    echo "  server.secretkey: $($pwmake 56 | base64)" >> $TMP_DIR/agent-managed/argocd-secret.yaml
+    echo "data:" >> $TMP_DIR/agent-autonomous/argocd-secret.yaml
+    echo "  server.secretkey: $($pwmake 56 | base64)" >> $TMP_DIR/agent-autonomous/argocd-secret.yaml
```

Comment:
```suggestion
    echo "data:" >> $TMP_DIR/agent-managed/argocd-secret.yaml
    echo "  server.secretkey: $($pwmake 56 | base64 -w0)" >> $TMP_DIR/agent-managed/argocd-secret.yaml
    echo "data:" >> $TMP_DIR/agent-autonomous/argocd-secret.yaml
    echo "  server.secretkey: $($pwmake 56 | base64 -w0)" >> $TMP_DIR/agent-autonomous/argocd-secret.yaml
```

Reply:
-w0 is to not add line wraps to base64 encoded output.

Reply:
Seems that macOS base64 doesn't break lines by default, and doesn't support the `-w0` option.  Have to add some logic to detect if we're on mac and tweak the command accordingly.

Reply:
The wonderful world of BSD vs. GNU :)

Reply:
Maybe we could just use `base64 "$inputfile" | tr -d '\n'`?

Reply:
I think detecting the OS is cleaner and clearer